app/src/main/database/index.ts

```diff
@@ -2,93 +2,262 @@ import fs from "fs";
+    }
 
-export const runMigrations = () => {
-  if (!databaseUrl) {
-    throw new Error(
-      "Database URL is not set. Ensure the database is opened first.",
-    );
+    return index;
   }
 
-  // Determine if we're in a packaged app or development
-  const isPackaged = __dirname.includes("app.asar");
-
-  let dbmatePath: string;
-  let migrationsDir: string;
-
-  if (isPackaged) {
-    // For packaged apps, binaries and migrations are outside the asar archive
-    dbmatePath = path.join(process.resourcesPath, "bin", "dbmate");
-    migrationsDir = path.join(process.resourcesPath, "migrations");
-  } else {
-    // In development, use paths relative to current working directory
-    dbmatePath = path.join(process.cwd(), "node_modules", ".bin", "dbmate");
-    migrationsDir = path.join(
-      process.cwd(),
-      "src",
-      "main",
-      "database",
-      "migrations",
-    );
+  getSourceItemVersions(sourceID: string): ItemVersions | null {
+    const itemVersions: ItemVersions = {};
+    for (const row of this.selectItemVersionsBySource.iterate({
+      sourceID: sourceID,
+    })) {
+      itemVersions[row.id] = row.version;
+    }
+    return itemVersions;
   }
 
-  // Ensure migrations directory exists
-  if (!fs.existsSync(migrationsDir)) {
-    throw new Error(`Migrations directory not found: ${migrationsDir}`);
+  // Updates the index version: should be called on any write operation to
+  // sources or items
+  private updateVersion() {
+    const index = this.getIndex();
+    const strIndex = JSON.stringify(index, sortKeys);
+    const newVersion = computeVersion(strIndex);
+    const replaceVersion = this.db?.transaction((version) => {
+      this.deleteVersion.run();
+      this.insertVersion.run(version);
+    });
+    if (replaceVersion) {
+      replaceVersion(newVersion);
+    }
   }
 
-  // Ensure dbmate binary exists
-  if (!fs.existsSync(dbmatePath)) {
-    throw new Error(`dbmate binary not found: ${dbmatePath}`);
+  deleteItems(items: string[]) {
+    for (const itemID in items) {
+      this.deleteItem.run({ id: itemID });
+    }
+    this.updateVersion();
   }
 
-  try {
-    const command = [
-      `"${dbmatePath}"`,
-      `--url "${databaseUrl}"`,
-      `--migrations-dir "${migrationsDir}"`,
-      // Don't update the schema file when running migrations at app startup
-      `--schema-file "/dev/null"`,
-      "up",
-    ].join(" ");
-
-    console.log("Running migrations:", command);
-    execSync(command, { stdio: "inherit" });
-    console.log("Migrations completed successfully");
-  } catch (error) {
-    console.error("Migration failed:", error);
-    throw new Error(`Migration failed: ${error}`);
+  deleteSources(sources: string[]) {
+    for (const sourceID in sources) {
+      this.deleteSource.run({ id: sourceID });
+      for (const itemRow of this.selectItemVersionsBySource.iterate({
+        sourceID: sourceID,
+      })) {
+        this.deleteItem.run({ id: itemRow.id });
+      }
```

Comment:
Until and unless `item.source_id` becomes a foreign key that we can `DELETE CASCADE` across, I think we should reverse these operations for internal consistency:


```suggestion
      for (const itemRow of this.selectItemVersionsBySource.iterate({
        sourceID: sourceID,
      })) {
        this.deleteItem.run({ id: itemRow.id });
      }
      this.deleteSource.run({ id: sourceID });
```

(There will indeed be cases where `deleteItems()` is called for a source, including over its entire collection, but the source itself remains.)

Reply:
good callout! reversed the order of the deletes here, and also moved this (+ the other update functions) to run within a transaction